src/Types/TypedValidator.purs

```diff
@@ -0,0 +1,149 @@
+
+-- | The type of the redeemers of this connection type.
+class RedeemerType (a :: Type) (b :: Type) | a -> b
+
+instance RedeemerType Void Void
+
+else instance RedeemerType Any PlutusData
+
+-- | Default instance
+else instance RedeemerType a Unit
+
+-- Replace `ScriptContext` by `Transaction` which contains all the scripts
+-- anyway:
+-- | The type of validators for the given connection type.
+type ValidatorType (a :: Type) (b :: Type) =
+  DatumType a b => RedeemerType a b => b -> b -> Transaction -> Boolean
+
+type WrappedValidatorType =
+  PlutusData -> PlutusData -> PlutusData -> Effect Unit
+
+wrapValidator
+  :: forall (d :: Type) (r :: Type)
+   . FromData d
+  => FromData r
+  => (d -> r -> Transaction -> Boolean)
+  -> WrappedValidatorType
+wrapValidator f d r p =
+  map (const unit) $ fromJustEff "Failed wrapValidator" $
+    f <$> fromData d <*> fromData r <*> fromData p
+
+-- Plutus rev: cc72a56eafb02333c96f662581b57504f8f8992f via Plutus-apps (localhost): abe4785a4fc4a10ba0c4e6417f0ab9f1b4169b26
+-- | A typed validator script with its `ValidatorScript` and `Address`.
+newtype TypedValidator (a :: Type) = TypedValidator
+  { validator :: Validator
+  , validatorHash :: ValidatorHash
+  , forwardingMPS :: MintingPolicy
+  , forwardingMPSHash :: MintingPolicyHash
+  -- The hash of the minting policy that checks whether the validator
+  -- is run in this transaction
+  }
+
+derive instance Generic (TypedValidator a) _
+derive instance Newtype (TypedValidator a) _
+derive newtype instance Eq (TypedValidator a)
+
+instance Show (TypedValidator a) where
+  show = genericShow
+
+-- Not sure how necessary this is:
+-- | Generalise the typed validator to one that works with the `PlutusData` type.
+generalise :: forall (a :: Type). TypedValidator a -> TypedValidator Any
+generalise
+  (TypedValidator { validator, validatorHash: vh, forwardingMPS, forwardingMPSHash }) =
+  -- we can do this safely because the on-chain validators are untyped, so they always
+  -- take 'PlutusData' arguments. The validator script stays the same, so the conversion
+  -- from 'PlutusData' to 'a' still takes place, even if it's not reflected in the type
+  -- signature anymore.
+  TypedValidator { validator, validatorHash: vh, forwardingMPS, forwardingMPSHash }
+
+-- | The hash of the validator.
+typedValidatorHash :: forall (a :: Type). TypedValidator a -> ValidatorHash
+typedValidatorHash = _.validatorHash <<< unwrap
+
+-- | The validator script itself.
+typedValidatorScript :: forall (a :: Type). TypedValidator a -> Validator
+typedValidatorScript = _.validator <<< unwrap
+
+-- | The minting policy that forwards all checks to the instance's validator.
+forwardingMintingPolicy
+  :: forall (a :: Type). TypedValidator a -> MintingPolicy
+forwardingMintingPolicy = _.forwardingMPS <<< unwrap
+
+-- We have a few functions, I'm not sure if we even need these for off chain
+-- code:
+-- -- Broken, see below (we need some notion of `applyCode`)
```

Comment:
We will have a server call that applies `PlutusData` arguments to a `Script` soon, but that is the only type of `applyCode` that we will be able to support (and it will be in `QueryM`)